### PR TITLE
[TP] Use Stride inferred from local tensor in to_local bwd

### PR DESCRIPTION
**Pull request by fduwjj**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #102630
* #102209

### CODE REVIEW

File: test/distributed/_tensor/test_dtensor.py [hunk 1]
```diff
@@ -230,6 +230,23 @@ def test_to_local(self):
 
         self.assertEqual(sharded_tensor.grad.to_local(), torch.ones(3, 3) * 3)
 
+        # test the case when grad stride is different from fwd input.
+        res = sharded_tensor.to_local()
+        model = torch.nn.ReLU()
+        res.register_hook(lambda grad: grad.t())
+        sharded_tensor.register_hook(
```
Comment (wanchaol) [inline test/distributed/_tensor/test_dtensor.py:237]: why this needs a hook registration, instead of just run the backward(), then compare `sharded_tensor.grad.stride(), [1, 12]`?

**Review by wanchaol — COMMENTED** [review]
Looks pretty good, I have a few suggestions inlined.

**Comment by wanchaol** [comment]
Also, it looks like the CI failure is related to the test

File: torch/distributed/_tensor/api.py [hunk 2]
```diff
@@ -58,14 +58,17 @@ def forward(ctx, input: "DTensor"):  # type: ignore[override]
...
         dtensor_spec = ctx.dtensor_spec
         dtensor_meta = dtensor_spec.tensor_meta
+        _, tensor_stride = _compute_global_tensor_info(
+            grad_output, dtensor_spec.mesh, dtensor_spec.placements
+        )
         return DTensor(
             grad_output,
             dtensor_spec.mesh,
             dtensor_spec.placements,
             shape=dtensor_meta.shape,
```
[the remark this answers is not in the document]
Comment (fduwjj) [reply torch/distributed/_tensor/api.py:68]: I thought about that too but the shape will always be same otherwise, the AutoGrad is not correct because grad will have the same size of input.

File: torch/distributed/_tensor/api.py [hunk 3]
```diff
@@ -536,3 +526,51 @@ def replicate_module_params_buffers(m: nn.Module, mesh: DeviceMesh) -> None:
...
+
+    Return:
+        tensor_shape: A List of int which specifies the size of DTensor which build
+            on top of the local tensor.
+        tensor_stride: A List of int which specifies the stride of DTensor.
+    """
+    tensor_shape = list(tensor.size())
+    tensor_stride = list(tensor.stride())
+    for idx, placement in enumerate(placements):
+        ws_dim_idx = mesh.size(idx)
```
[the remark this answers is not in the document]
Comment (fduwjj) [reply torch/distributed/_tensor/api.py:562]: world size dim per idx.. Well, mesh_dim is a better name.

File: test/distributed/_tensor/test_dtensor.py (hunk 1 above)
Comment (fduwjj) [reply test/distributed/_tensor/test_dtensor.py:237]: Originally the check failed in the end... Looks like it's working now. Let me remove it.

**Review by wanchaol — APPROVED** [review]
lgtm, thanks for the fix!

**Comment by fduwjj** [comment]
Fixed the import error and run `lintrunner` locally. It all passed.
